src/app/clusters/actions-server/actions-server.h

```diff
@@ -0,0 +1,353 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandlerInterface.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+namespace {
+// Zero'th index will be used for unit tests only.
+static constexpr size_t kActionsDelegateTableSize =
+#ifdef ZCL_USING_ACTIONS_CLUSTER_SERVER
+    MATTER_DM_ACTIONS_CLUSTER_SERVER_ENDPOINT_COUNT +
+#endif
+    CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT + 1;
+
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Actions {
+
+// Last byte is reserved of '\0' terminator.
+static constexpr size_t kActionNameMaxSize       = 127u;
+static constexpr size_t kEndpointListNameMaxSize = 127u;
+
+static constexpr size_t kEndpointListMaxSize = 256u;
+
+class Delegate;
+
+struct ActionStructStorage : public Structs::ActionStruct::Type
+{
+    ActionStructStorage() : mActionName(mBuffer, sizeof(mBuffer)){};
+
+    ActionStructStorage(uint16_t action, const CharSpan & actionName, ActionTypeEnum actionType, uint16_t epListID,
+                        BitMask<CommandBits> commands, ActionStateEnum actionState) :
+        mActionName(mBuffer, sizeof(mBuffer))
+    {
+        Set(action, actionName, actionType, epListID, commands, actionState);
+    }
+
+    ActionStructStorage(const ActionStructStorage & action) : mActionName(mBuffer, sizeof(mBuffer)) { *this = action; }
+
+    ActionStructStorage & operator=(const ActionStructStorage & action)
+    {
+        Set(action.actionID, action.name, action.type, action.endpointListID, action.supportedCommands, action.state);
+        return *this;
+    }
+
+    void Set(uint16_t action, const CharSpan & actionName, ActionTypeEnum actionType, uint16_t epListID,
+             BitMask<CommandBits> commands, ActionStateEnum actionState)
+    {
+        actionID          = action;
+        type              = actionType;
+        endpointListID    = epListID;
+        supportedCommands = commands;
+        state             = actionState;
+        CopyCharSpanToMutableCharSpanWithTruncation(actionName, mActionName);
```

Comment:
This will break if setting a longer name than the last one we set, right?

Reply:
Reset the span with the actual buffer before copying.